plugin/nomad/nomad.go

```diff
@@ -0,0 +1,156 @@
+package nomad
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/metrics"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/hashicorp/nomad/api"
+	"github.com/miekg/dns"
+)
+
+const pluginName = "nomad"
+
+var (
+	log        = clog.NewWithPlugin(pluginName)
+	defaultTTL = 30
+)
+
+type Nomad struct {
+	Next plugin.Handler
+
+	ttl     uint32
+	Zone    string
+	clients []*api.Client
+	current int
+}
+
+func (n *Nomad) Name() string {
+	return pluginName
+}
+
+func (n Nomad) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	qname, originalQName := processQName(state.Name(), n.Zone)
+
+	namespace, serviceName, err := extractNamespaceAndService(qname)
+	if err != nil {
+		return plugin.NextOrFailure(n.Name(), n.Next, ctx, w, r)
+	}
+
+	m, header := initializeMessage(state, n.ttl)
+
+	svcRegistrations, _, err := fetchServiceRegistrations(n, serviceName, namespace)
+	if err != nil {
+		log.Warning(err)
+		return handleServiceLookupError(w, m, ctx, namespace)
+	}
+
+	if len(svcRegistrations) == 0 {
+		return handleResponseError(n, w, m, originalQName, n.ttl, ctx, namespace, err)
+	}
+
+	if err := addServiceResponses(m, svcRegistrations, header, state.QType(), originalQName, n.ttl); err != nil {
+		return handleResponseError(n, w, m, originalQName, n.ttl, ctx, namespace, err)
+	}
+
+	err = w.WriteMsg(m)
+	requestSuccessCount.WithLabelValues(metrics.WithServer(ctx), namespace).Inc()
+	return dns.RcodeSuccess, err
+}
+
+func processQName(qname string, zone string) (string, string) {
+	originalQName := qname
+	qname = strings.ReplaceAll(qname, zone, "")
+	qname = strings.Trim(qname, ".")
+	return qname, originalQName
+}
```

Comment:
For the sake of robustness, we could use a helper function from `github.com/coredns/coredns/plugin/pkg/dnsutil` here instead, like:

```suggestion
func processQName(qname, zone string) (string, string, error) {
	original := dns.Fqdn(qname)
	base, _ := dnsutil.TrimZone(original, dns.Fqdn(zone)) // but with err handling
	return base, original
}
```